router/main_test.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+	"time"
+)
+
+func Test_makeHandler(t *testing.T) {
+	config := []struct {
+		RequestHost        string
+		UpstreamURL        string
+		ExpectedStatusCode int
+	}{
+		{
+			RequestHost:        "martindekov.example.xyz/",
+			UpstreamURL:        "http://localhost:8080/myfunction",
+			ExpectedStatusCode: http.StatusInternalServerError,
+		},
+		{
+			RequestHost:        "martindekov.example.xyz/",
+			UpstreamURL:        "/myfunction",
+			ExpectedStatusCode: http.StatusServiceUnavailable,
+		},
+		{
+			RequestHost:        "",
+			UpstreamURL:        "http://localhost:8080/myfunction",
+			ExpectedStatusCode: http.StatusBadRequest,
+		},
+		{
+			RequestHost:        "",
+			UpstreamURL:        "http://localhost:8080/",
+			ExpectedStatusCode: http.StatusBadRequest,
+		},
+	}
+
+	timeout := time.Second * 30
+
+	for _, test := range config {
+		req, err := http.NewRequest(http.MethodPost, "/", nil)
+		if err != nil {
+			t.Fatalf("Request error: `%v`\n", err)
+		}
+
+		URI := strings.TrimPrefix(test.UpstreamURL, "http://localhost:8080")
+		req.RequestURI = URI
+		req.Host = test.RequestHost
+
+		rec := httptest.NewRecorder()
+		c := &http.Client{
+			Timeout: timeout,
+		}
+
+		URL := strings.TrimSuffix(test.UpstreamURL, "myfunction")
+		handlerFunc := makeHandler(c, timeout, URL)
+		handlerFunc(rec, req)
+
+		if rec.Code != test.ExpectedStatusCode {
+			t.Errorf("Expected status code : `%v` got: `%v`", test.ExpectedStatusCode, rec.Code)
+		}
+
+		read, err := ioutil.ReadAll(rec.Body)
+		if err != nil {
+			t.Errorf("Unexpected response body read error: `%v`", err)
+		}
+
+		bodyValue := string(read)
+		var ExpectedSuffix string
+		userFromURI := strings.TrimPrefix(URI, "/")
```

Comment:
Seems like we're doing a lot of trimming and checking in the test. This makes it look like we're testing the wrong thing(s) or the wrong level of granularity. Any thoughts?

Normally a test is:

Arrange - some data

Act - call the method passing in the data

Assert - check that the actual vs expected match.

Reply:
You requested to reduce the number of fields in struct to max 3 that's why I did so much trimming etc. Should I increment the test cases again?